can/interfaces/ixxat/canlib_vcinpl2.py

```diff
@@ -0,0 +1,920 @@
+        pBtpFDR = constants.CAN_DATABITRATE_PRESETS[data_bitrate]
+        log.info(
+            "Initializing control %d with SDR={%s}, FDR={%s}",
+            channel,
+            pBtpSDR,
+            pBtpFDR,
+        )
+        _canlib.canControlOpen(
+            self._device_handle, channel, ctypes.byref(self._control_handle)
+        )
+
+        _canlib.canControlGetCaps(
+            self._control_handle, ctypes.byref(self._channel_capabilities)
+        )
+
+        # check capabilities
+        bOpMode = constants.CAN_OPMODE_UNDEFINED
+        if (
+            self._channel_capabilities.dwFeatures & constants.CAN_FEATURE_STDANDEXT
+        ) != 0:
+            # controller supportes CAN_OPMODE_STANDARD and CAN_OPMODE_EXTENDED at the same time
+            bOpMode |= constants.CAN_OPMODE_STANDARD  # enable both 11 bits reception
+            if extended:  # parameter from configuration
+                bOpMode |= constants.CAN_OPMODE_EXTENDED  # enable 29 bits reception
+        elif (
+            self._channel_capabilities.dwFeatures & constants.CAN_FEATURE_STDANDEXT
+        ) != 0:
+            log.warning(
+                "Channel %d capabilities allow either basic or extended IDs, but not both. using %s according to parameter [extended=%s]",
+                channel,
+                "extended" if extended else "basic",
+                "True" if extended else "False",
+            )
+            bOpMode |= (
+                constants.CAN_OPMODE_EXTENDED
+                if extended
+                else constants.CAN_OPMODE_STANDARD
+            )
+
+        if (
+            self._channel_capabilities.dwFeatures & constants.CAN_FEATURE_ERRFRAME
+        ) != 0:
+            bOpMode |= constants.CAN_OPMODE_ERRFRAME
+
+        bExMode = constants.CAN_EXMODE_DISABLED
+        if (self._channel_capabilities.dwFeatures & constants.CAN_FEATURE_EXTDATA) != 0:
+            bExMode |= constants.CAN_EXMODE_EXTDATALEN
+
+        if (
+            self._channel_capabilities.dwFeatures & constants.CAN_FEATURE_FASTDATA
+        ) != 0:
+            bExMode |= constants.CAN_EXMODE_FASTDATA
+
+        _canlib.canControlInitialize(
+            self._control_handle,
+            bOpMode,
+            bExMode,
+            constants.CAN_FILTER_PASS,
+            constants.CAN_FILTER_PASS,
+            0,
+            0,
+            ctypes.byref(pBtpSDR),
+            ctypes.byref(pBtpFDR),
+        )
+
+        # With receive messages, this field contains the relative reception time of
+        # the message in ticks. The resolution of a tick can be calculated from the fields
+        # dwClockFreq and dwTscDivisor of the structure  CANCAPABILITIES in accordance with the following formula:
+        # frequency [1/s] = dwClockFreq / dwTscDivisor
+        # We explicitly cast to float for Python 2.x users
+        self._tick_resolution = float(
+            self._channel_capabilities.dwTscClkFreq  # TODO confirm
```

Comment:
Think this looks OK, maybe remove the TODO.

Reply:
Removed